server/cluster.go

```diff
@@ -0,0 +1,245 @@
+// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package server
+
+import (
+	"bytes"
+	"encoding/hex"
+	"fmt"
+	"sync"
+
+	"github.com/golang/protobuf/proto"
+	"github.com/pingcap/kvproto/pkg/metapb"
+	"github.com/pingcap/kvproto/pkg/pdpb"
+	"github.com/pkg/errors"
+	log "github.com/sirupsen/logrus"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+const defaultTiKVVersion = "3.0.0"
+
+var (
+	emptyKey = []byte{}
+)
+
+type clusterInstance struct {
+	sync.RWMutex
+	server       *Server
+	clusterID    uint64
+	maxPeerCount uint32
+	safePoint    uint64
+
+	member     *memberInstance // we only support one leader pd at this time
+	regions    []*regionInstance
+	regionByID map[uint64]*regionInstance
+	stores     []*storeInstance
+	storeByID  map[uint64]*storeInstance
+
+	idAlloc uint64
+
+	store *MVCCLevelDB
+}
+
+func validateRegionAndStore(regions []*regionInstance, stores []*storeInstance) error {
+	var lastEndKey []byte = nil
+	var lastRegion *regionInstance = nil
+	for _, region := range regions {
+		if lastEndKey == nil {
+			if len(region.StartKey) != 0 {
+				return errors.New(fmt.Sprintf("Start key of first region '%d' isn't empty", region.Id))
+			}
+			if region.EndKey == nil {
+				region.EndKey = emptyKey
+			}
+		} else {
+			if !bytes.Equal(region.StartKey, lastEndKey) {
+				return errors.New(fmt.Sprintf("Start key '%s' of region '%d' isn't the same as end key '%s' of last region '%d'",
+					hex.EncodeToString(region.StartKey), region.Id,
+					hex.EncodeToString(lastEndKey), lastRegion.Id,
+				))
+			}
+		}
+		if len(region.EndKey) > 0 && len(region.StartKey) > 0 && bytes.Compare(region.StartKey, region.EndKey) >= 0 {
+			return errors.New(fmt.Sprintf("Start key '%s' of region '%d' isn't the less than end key '%s'",
+				hex.EncodeToString(region.StartKey), region.Id,
+				hex.EncodeToString(region.EndKey),
+			))
+		}
+
+		lastRegion = region
+		lastEndKey = region.EndKey
+	}
+	if len(lastEndKey) != 0 {
+		return errors.New(fmt.Sprintf("Last end key '%s' of region '%d' isn't empty",
+			hex.EncodeToString(lastEndKey), lastRegion.Id,
+		))
+	}
+	return nil
+}
+
+func (c *clusterInstance) startStores(address string) error {
+	for _, store := range c.stores {
```

Comment:
Read lock?

Reply:
The connection information isn't returned therefore no client can connect to it at this time.